### DISCUSSION TITLE
Is there any way to conditionally disable a nodes execution?

### DISCUSSION THREAD

**Discussion by shawnington** [Comfy-Org/ComfyUI#3639] [General]
Im very very close to having a workflow that takes 1.5 or SDXL models, and switches all the control nets and other things to the appropriate equivalents for each model, the only thing I still have to do manually is bypass the SDXL conditioning, as the 1.5 model doesn't have the second text encoder, and it freaks out if the conditioning is applied even though the output isn't used. 

I wrote a custom node that tells what the model type is, that information is connected to switches that swap things. 

I could write a custom node for the SDXL text prompt conditioning that has an enable disable input, but I was wondering if there is another way to conditionally enable/disable the node? 

Is the onTrigger mode actually implemented? I see it in the litegraph.js but, how do you trigger an event? Seems like a solution. I already have a custom node that emits model type which is used to switch things like 1.5 open pose to sdxl open pose, so if the OnEvent or OnTrigger modes actually work, what format of input do they take?

@ltdrdata 

<img width="1012" alt="Screenshot 2024-06-03 at 5 45 54 PM" src="https://github.com/comfyanonymous/ComfyUI/assets/88048838/b2654f38-58a9-45e2-81e7-83a7654f35e2">

**Comment by ltdrdata** [comment]
OnEvent, OnTrigger, etc., only function on the front end and are determined before the workflow is executed. In ComfyUI, this corresponds to setting the mute state.

The approach depends entirely on whether you want a switch that is applied before execution or runtime control flow.

**Reply by shawnington** [reply]
I have a node that says, this is a 1.5 model or an SDXL model, and I want to be able to say hey SDXL encode dont run, Im a 1.5 model, or hey SDXL encode, run because Im an SDXL model. It will throw an error if not muted, so runtime control flow. Run time, say hey, Im running a 1.5 model, don't run please becuase I don't have the 2nd text encoder.

Any guidance you can give me in where In the code base I should look for reference as to how to do this, or an example. Id be grateful. I defer to you as the node wizard.

**Reply by sttollgrin** [reply]
I was looking for something similar (for programmatically bypassing whole groups of logic, ie.: conditionally bypassing upscale/hand/face refiner etc) and eventually upon stumbling on  https://docs.comfy.org/essentials/custom_node_lazy_evaluation#lazy-evaluation I came up with this:
```python
class AnyType(str):
    def __ne__(self, __value: object) -> bool:
        return False


any_type = AnyType("*")


class Bypasser:

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "input_select": ("INT", {"defaultInput": True},),
                "input1": (any_type, {"defaultInput": True, "lazy": True},),
                "input2": (any_type, {"defaultInput": True, "lazy": True},),
            }
        }

    RETURN_TYPES = (any_type,)
    FUNCTION = "switch"
    CATEGORY = "Bypass-Support"

    def check_lazy_status(self, input_select, input1=None, input2=None):
        needed = []
        if input_select == 1:
            needed.append("input1")
        elif input_select == 2:
            needed.append("input2")
        return needed

    def switch(self, input_select, input1=None, input2=None):
        result = None
        if input_select == 1:
            result = input1
        elif input_select == 2:
            result = input2

        return (result,)

```

this node will conditionally execute only part of the execution tree with nodes connected to either input1 or input2

here's a minimum example of this:
![workflow (2)](https://github.com/user-attachments/assets/a9e1148e-7eff-4171-84e0-422630b08085)
here with non-SDXL model and SDXL nodes are skipped

**Reply by woodrowbarlow** [reply]
thank you; this helped me grep around for an existing node.

for users of "impactpack", this seems to be exactly what "ImpactConditionalBranch" does --

https://github.com/ltdrdata/ComfyUI-Impact-Pack/blob/1c488eec11264db23821c864f1482a99ba9893d0/modules/impact/logics.py#L78

**Reply by schoenid** [reply]
> I have a node that says, this is a 1.5 model or an SDXL model

@shawnington Is this an official node? 
I was searching for something to differentiate SD1.5, SDXL, FLUX, but didn't find anything.
So I've renamed all my models to get it in the filename and created a workflow to filter it out ...
I would be happy, if there would exist an easier way.

**Comment by ltdrdata** [comment]
This workflow can now be configured through Impact Pack.
(The latest version of ComfyUI is required)

https://github.com/ltdrdata/ComfyUI-extension-tutorials/blob/Main/ComfyUI-Impact-Pack/tutorial/switch.md

**Comment by ArtHommage** [comment]
What I want is for an entire group of nodes to not execute based on a condition evaluated early running a job. For instance, to check an image loaded for if the resolution is over or under a certain margin, and to skip executing a group that contains, say, captioning or upscaling (in the way bypass would). I can logically switch data flow, but I'd rather not execute certain nodes at all.

**Reply by ltdrdata** [reply]
> What I want is for an entire group of nodes to not execute based on a condition evaluated early running a job. For instance, to check an image loaded for if the resolution is over or under a certain margin, and to skip executing a group that contains, say, captioning or upscaling (in the way bypass would). I can logically switch data flow, but I'd rather not execute certain nodes at all.

Using the switch nodes in Impact Pack completely skips the execution of specific nodes.

**Reply by t00350320** [reply]
Imagine there are two inputs:input1 and input2, input1 equals model1 amd input2 equas model2.
Even we select input2 to be OUTPUT,  the input1's models load and preparation always done which supposed should do nothing for input1's action. It means that, if we have a method to cut off input1's connection in Swicth Node?

**Reply by ltdrdata** [reply]
> Imagine there are two inputs:input1 and input2, input1 equals model1 amd input2 equas model2.
> Even we select input2 to be OUTPUT,  the input1's models load and preparation always done which supposed should do nothing for input1's action. It means that, if we have a method to cut off input1's connection in Swicth Node?

You can use `Impact Switch` node with `select_on_prompt` mode. It simulates disconnecting input before executing worklfow based on selection.

**Reply by t00350320** [reply]
![0559D581-9407-493E-A88F-021C6E6AA7AB](https://github.com/user-attachments/assets/73c00eb8-920d-449f-ad68-d2e0ffad5e60)

want to choose differ ipa models by different    selection, always run both two ipa models even select input2. have no idea why it can not work? please help to check? thanks

**Reply by ltdrdata** [reply]
> ![0559D581-9407-493E-A88F-021C6E6AA7AB](https://github.com/user-attachments/assets/73c00eb8-920d-449f-ad68-d2e0ffad5e60)
> 
> want to choose differ ipa models by different    selection, always run both two ipa models even select input2. have no idea why it can not work? please help to check? thanks 

`select_on_prompt` does not allow the select input to be determined at runtime.

In your case, instead of using `Switch`, you need to use `Inversed Switch` with `select_on_execution`. This will allow you to switch the model going into IPAdapter Advanced.

**Reply by t00350320** [reply]
Thank you!!! 
It works, put "Inversed Switch" Node before two IPA nodes , and work with  current "Switch(Any)" Node.
By the way, IPA input parameter "model"  will not be empty, so hack this code to skip model check, this works well  with "Inversed Switch"

**Reply by kris775** [reply]
@t00350320 , I am also badly in need of a similar workflow to bypass IPA during some stages, my hack is not working, can you please share the hack you did for the IPA Advanced node. Thank you.

**Reply by t00350320** [reply]
> @t00350320 , I am also badly in need of a similar workflow to bypass IPA during some stages, my hack is not working, can you please share the hack you did for the IPA Advanced node. Thank you.
```
In  ComfyUI_IPAdapter_plus/IPAdapterPlus.py ,  
class IPAdapterAdvanced:
xxx
  def apply_ipadapter(self, model, ipadapter, xxx):
        if model==None:   【add this line】
            return(None,None,)  【add this line】
        is_sdxl = isinstance

**Reply by kris775** [reply]
@t00350320 Thank you, That works. In your case where you are  putting  "Inversed Switch" Node before two IPA nodes , and work with current "Switch(Any)" Node after that to ksampler , for "Switch(Any)" Node are you using select_on_prompt mode or select_on_execution mode.
@ltdrdata  For me "inverse switch "is working perfectly by guiding checkpoint directly as output 1 and to IPAdapter as output2 with "select_on_execution" and  i can control the execution path of the workflow dynamically. But the subsequent selection of these two streams of input through "Switch(Any)" Node is not working with select_on_execution mode, works only with select_on_prompt mode and hence cant control it dynmically. Any Help With this, Thank you.

**Reply by ltdrdata** [reply]
Have you read this tutorial?  
https://github.com/ltdrdata/ComfyUI-extension-tutorials/blob/Main/ComfyUI-Impact-Pack/tutorial/switch.md
You should also carefully read the precautions there.

**Reply by ArtHommage** [reply]
I hate to be daft, but I think I'm still missing something, or not being understood. Here is my specific situaton:
I have a BBOX detection that creates a mask for a specific region. Under that is a detailer node (Specifically "FaceDetailer") that I only want to run when a mask is detected.  The issue I am having is that just choking off the image input does not prevent the detailer node from running, and it errors.

![image](https://github.com/user-attachments/assets/dfc957f1-8112-4f24-a7ac-a6e0360cad07)

[Disregard the custom nodes I'm playing with]

I'm certain there's a much more elegant way to do this, likely with putting them in a group, but I'm not sure how to programmatically mute a group during execution, or to just skip it.

**Comment by schoenid** [comment]
@ArtHommage 

> I have a BBOX detection that creates a mask for a specific region. Under that is a detailer node (Specifically "FaceDetailer") that I only want to run when a mask is detected.

Could a Control Bridge (Impact-Pack) solve your problem?

![image](https://github.com/user-attachments/assets/f58481a4-d4a8-484a-a77d-b682a45ed93d)

I've used it, to detect wether a checkpoint has CLIP embedded ... it turns off the following node in this example.

**Comment by schoenid** [comment]
> [schoenid](https://github.com/schoenid)
> [on Oct 2, 2024](https://github.com/comfyanonymous/ComfyUI/discussions/3639#discussioncomment-10814951)
> I have a node that says, this is a 1.5 model or an SDXL model
> 
> @shawnington Is this an official node?
> I was searching for something to differentiate SD1.5, SDXL, FLUX, but didn't find anything.
> So I've renamed all my models to get it in the filename and created a workflow to filter it out ...
> I would be happy, if there would exist an easier way.

So I found a solution and created my own node:

![image](https://github.com/user-attachments/assets/22243d13-89cd-414a-b48f-92567467d9df)

Code:

```
class get_ModelType:
    def __init__(self):
        pass

    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "model": ("MODEL",),
                "output": ("STRING", {
                    "multiline": True,
                    "default": ""
                }),
            },
        }

    RETURN_TYPES = ("STRING",)
    RETURN_NAMES = ("Model Type",)
    FUNCTION = "model_check"
    OUTPUT_NODE = True
    CATEGORY = "my_custom"

    def model_check(self, output, model=None):
        results = ""
        modeltype = ""
        if model:
            modeltype = "Base:\t\t" + model.model.latent_format.__class__.__name__ + "\n"
            results += modeltype

            dtype_value = model.model.get_dtype()  # eg float8
            dtype_str = str(dtype_value).split('.')[-1]  # get part after '.'
            print(dtype_str)

            results += f"dtype:\t\t{dtype_str}\n"
            results += f"Model Kind:\t{model.model.model_type.name}\n"
            results += "\n"

        modeltype = modeltype.split(':')[-1].strip()

        return {"ui": {"output": results}, "result": (modeltype,)}


NODE_CLASS_MAPPINGS = {
    "getModelType": get_ModelType
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "getModelType": "get ModelType"
}

```
